src/cunumeric/random/bitgenerator_curand.inl

```diff
@@ -0,0 +1,343 @@
+      gen.skip_ahead(initialoffset + totalcount - gen.offset);
+
+    return totalcount;
+  }
+};
+
+template <VariantKind kind>
+struct generatormap {
+  generatormap() {}
+  ~generatormap()
+  {
+    std::lock_guard<std::mutex> guard(lock);
+    if (m_generators.size() != 0) {
+      ::fprintf(stderr, "[ERROR] : some generators have not been freed - LEAK !\n");
+      // TODO: assert ?
+    }
+  }
+
+  std::mutex lock;
+  std::map<int, std::unique_ptr<CURANDGenerator>> m_generators;
+
+  bool has(int32_t generatorID)
+  {
+    std::lock_guard<std::mutex> guard(lock);
+    return m_generators.find(generatorID) != m_generators.end();
+  }
+
+  CURANDGenerator* get(int32_t generatorID)
+  {
+    std::lock_guard<std::mutex> guard(lock);
+    if (m_generators.find(generatorID) == m_generators.end()) {
+      ::fprintf(stderr,
+                "[ERROR] : internal error : generator ID <%d> does not exist (destroy) !\n",
+                generatorID);
+      assert(false);
+    }
+    return m_generators[generatorID].get();
+  }
+
+  void create(int32_t generatorID, BitGeneratorType gentype)
+  {
+    CURANDGenerator* cugenptr = CURANDGeneratorBuilder<kind>::build(gentype);
+
+    std::lock_guard<std::mutex> guard(lock);
+    // safety check
+    if (m_generators.find(generatorID) != m_generators.end()) {
+      ::fprintf(
+        stderr, "[ERROR] : internal error : generator ID <%d> already in use !\n", generatorID);
+      assert(false);
+    }
+    m_generators[generatorID] = std::move(std::unique_ptr<CURANDGenerator>(cugenptr));
+  }
+
+  void destroy(int32_t generatorID)
+  {
+    std::unique_ptr<CURANDGenerator> cugenptr;
+    // verify it existed, and otherwise remove it from list
+    {
+      std::lock_guard<std::mutex> guard(lock);
+      if (m_generators.find(generatorID) == m_generators.end()) {
+        ::fprintf(stderr,
+                  "[ERROR] : internal error : generator ID <%d> does not exist (destroy) !\n",
+                  generatorID);
+        assert(false);
+      }
+      cugenptr = std::move(m_generators[generatorID]);
+      m_generators.erase(generatorID);
+    }
+
+    CURANDGeneratorBuilder<kind>::destroy(cugenptr.get());
+  }
+
+  void set_seed(int32_t generatorID, uint64_t seed)
+  {
+    CURANDGenerator* genptr = get(generatorID);
+    std::lock_guard<std::mutex> guard(genptr->lock);
+    CHECK_CURAND(::curandSetPseudoRandomGeneratorSeed(genptr->gen, seed));
+  }
+};
+
+template <VariantKind kind>
+struct BitGeneratorImplBody {
+  using generatormap_t = generatormap<kind>;
+
+  static std::mutex lock_generators;
+  static std::map<Legion::Processor, std::unique_ptr<generatormap_t>> m_generators;
+
+ private:
+  static generatormap_t& getgenmap()
+  {
+    const auto proc = Legion::Processor::get_executing_processor();
+    lock_generators.lock();
+    if (m_generators.find(proc) == m_generators.end()) {
+      m_generators[proc] = std::move(std::unique_ptr<generatormap_t>(new generatormap_t()));
```

Comment:
first of all, r-value doesn't need to be moved. second, please use `make_unique` instead of a manual construction. Manual `unique_ptr` constructions have their own uses and this isn't one of them.